equinox/_module.py

```diff
@@ -904,19 +850,86 @@ def __repr__(self):
         return repr((self.dynamic_field_names, self.static_fields))[1:-1]
 
 
-def _unflatten_module(cls: type["Module"], aux: _FlattenedData, dynamic_field_values):
-    # This doesn't go via `__init__`. A user may have done something nontrivial there,
-    # and the field values may be dummy values as used in various places throughout JAX.
-    # See also
-    # https://jax.readthedocs.io/en/latest/pytrees.html#custom-pytrees-and-initialization,
-    # which was (I believe) inspired by Equinox's approach here.
-    module = object.__new__(cls)
-    for name, value in zip(aux.dynamic_field_names, dynamic_field_values):
-        object.__setattr__(module, name, value)
-    for name, value in itertools.chain(aux.static_fields, aux.wrapper_fields):
-        if value is not _flatten_sentinel:
+class ModuleFlattener:
+    """Manager for (un)flattening `Module`."""
+
+    __slots__: tuple[str, str] = ("dynamic_fs", "static_fs")
+    dynamic_fs: tuple[str, ...]
+    static_fs: tuple[str, ...]
+
+    def __init__(self, fields: tuple[dataclasses.Field[Any], ...]) -> None:
+        self.dynamic_fs = tuple(
+            [f.name for f in fields if not f.metadata.get("static", False)]
+        )
+        self.static_fs = tuple(
+            [f.name for f in fields if f.metadata.get("static", False)]
+        )
+
+    def flatten(self, obj: "Module", /) -> tuple[tuple[PyTree, ...], _FlattenedData]:
+        """Flattens a module into a flattened representation."""
+        d = obj.__dict__
+
+        # Build dynamic keys, values in a single pass
+        dynamic_fs = []
+        dynamic_vs = []
+        for k in self.dynamic_fs:
+            v = d.get(k, _flatten_sentinel)
```

Comment:
Btw, does replacing this with
```
get = obj.__dict__.get
...
v = get(k, _flatten_sentinel)
```
help performance here at all? (If only by some very minor amount.)

Reply:
It should...

Reply:
Done.

Reply:
🤷 hard to tell what diff fit made. Well within the run-to-run differences for `%timeit`. Hopefully #1001 will help and I can go back and A/B test this more rigorously.